src/main/org/deidentifier/arx/algorithm/ExponentialMechanism.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * ARX: Powerful Data Anonymization
+ * Copyright 2012 - 2017 Fabian Prasser, Florian Kohlmayer and contributors
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.deidentifier.arx.algorithm;
+
+import java.math.BigDecimal;
+import java.math.MathContext;
+import java.math.RoundingMode;
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Random;
+
+import org.apache.commons.math3.analysis.function.Exp;
+import org.apache.commons.math3.distribution.EnumeratedDistribution;
+import org.apache.commons.math3.random.AbstractRandomGenerator;
+import org.apache.commons.math3.util.Pair;
+
+/**
+ * An implementation of the exponential mechanism for discrete domains as proposed in:
+ * 
+ * McSherry, Frank, and Kunal Talwar:
+ * Mechanism design via differential privacy.
+ * Foundations of Computer Science 2007. pp. 94-103
+ * 
+ * @author Raffael Bild
+ */
+public class ExponentialMechanism<T> {
+
+    /** The precision to use for BigDecimal arithmetic */
+    public static final int defaultPrecision = 100;
+
+    /** The math context to use for BigDecimal arithmetic */
+    private MathContext     mc;
+
+    /** A cryptographically strong random generator */
+    private static class SecureRandomGenerator extends AbstractRandomGenerator {

```

Comment:
This is necessary in order to use the random number generator with the apache commons math framework